thetis/adr_model.py

```diff
@@ -0,0 +1,220 @@
+import firedrake as fd
+import networkx as nx
+import sympy
+from sympy.parsing.sympy_parser import \
+    convert_xor, \
+    parse_expr, \
+    standard_transformations
+from sympy.utilities.lambdify import lambdify
+
+
+class ADR_Model:
+    """
+    Data structure to store an ADR model specification.
+
+    Performs validation checks on a user-provided dictionary of strings
+    and parses strings representing mathematical expressions into
+    SymPy expressions.
+
+    :cvar transformations: Transformations applied to strings parsed as
+        SymPy expressions.
+    :type transformations: tuple
+    :ivar constants:  Dictionary mapping strings used to uniquely identify
+        constants to the values of those constants.
+    :type constants: dict
+    :ivar species: Dictionary mapping strings used to uniquely identify
+        biological or chemical species to dictionaries that contain data
+        pertaining to those species.
+    :type species: dict
+    """
+
+    # convert_xor causes "x^2" to be parsed as "x**2"
+    transformations = (
+        standard_transformations + (convert_xor,))
+
+    reserved_symbols = {
+        "pi", "e",
+        "x", "y", "z", "t",
+        "u", "u_x", "u_y", "u_z"}
```

Comment:
I guess a discussion point is whether we want all of these (and any more) as reserved symbols. I can imagine pi, e, x, y, t being useful for if you have a source term that depends on them. Probably not z in a 2D model.

Reply:
Agreed -- see separate comment. For z, I suppose the question is whether or not this class might also be used in tandem with 3D solvers.